llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.cpp

```diff
@@ -0,0 +1,861 @@
+    // the head of the group.
+    for (auto *U : G1.getUses()) {
+      assert(U != SpillInstruction);
+      if (U == Head)
+        continue;
+
+      U->substituteRegister(DefRegToSpill, Restore->getOperand(0).getReg(), 0,
+                            *TRI);
+      RestoreUses.push_back(U);
+      LLVM_DEBUG(dbgs() << "Updated use: " << *U);
+      LLVM_DEBUG(dbgs() << "With register = "
+                        << printReg(Restore->getOperand(0).getReg(), TRI)
+                        << "\n");
+    }
+  }
+}
+
+SetVectorType AMDGPUEarlyRegisterSpilling::collectUsesDominatedBySpill(
+    Register DefRegToSpill, MachineInstr *SpillInstruction,
+    const SetVectorType &UnreachableUses) {
+  SetVectorType DominatedUses;
+  MachineBasicBlock *SpillBlock = SpillInstruction->getParent();
+  for (MachineInstr &U : MRI->use_nodbg_instructions(DefRegToSpill)) {
+    if (&U == SpillInstruction)
+      continue;
+
+    if (UnreachableUses.contains(&U))
+      continue;
+
+    if (U.isPHI()) {
+      for (auto *PhiOpMBB : getPhiBlocksOfSpillReg(&U, DefRegToSpill)) {
+        if (DT->dominates(SpillBlock, PhiOpMBB)) {
+          DominatedUses.insert(&U);
+        }
+      }
+    } else if (DT->dominates(SpillInstruction, &U)) {
+      DominatedUses.insert(&U);
+    }
+  }
+  return DominatedUses;
+}
+
+void AMDGPUEarlyRegisterSpilling::emitRestores(
+    Register DefRegToSpill, MachineInstr *CurMI, MachineInstr *SpillInstruction,
+    const SetVectorType &UnreachableUses, const TargetRegisterClass *RC,
+    int FI) {
+  assert(MRI->hasOneDef(DefRegToSpill) &&
+         "The Register does not have one definition");
+  MachineInstr *InstrOfDefRegToSpill =
+      MRI->getOneDef(DefRegToSpill)->getParent();
+
+  // Collect the uses that are dominated by SpillInstruction
+  SetVectorType DominatedUses = collectUsesDominatedBySpill(
+      DefRegToSpill, SpillInstruction, UnreachableUses);
+
+  SmallVector<MachineInstr *> RestoreInstrs;
+  SmallVector<MachineInstr *> RestoreUses;
+  emitRestoreInstrsForDominatedUses(DefRegToSpill, SpillInstruction, CurMI,
+                                    DominatedUses, RestoreInstrs, RestoreUses,
+                                    FI);
+
+  // Update the live interval analysis.
+  updateIndexes(InstrOfDefRegToSpill);
+  updateIndexes(SpillInstruction);
+  updateLiveness(InstrOfDefRegToSpill);
+  updateLiveness(SpillInstruction);
+
+  if (InstrOfDefRegToSpill != CurMI) {
+    updateIndexes(CurMI);
+    updateLiveness(CurMI);
+  }
+
+  for (auto *Use : RestoreInstrs) {
+    updateIndexes(Use);
+    updateLiveness(Use);
+  }
+
+  for (auto *Use : RestoreUses) {
+    updateIndexes(Use);
+    updateLiveness(Use);
+  }
+}
+
+// We have to collect the unreachable uses before we emit the spill instruction.
+// This is due to the fact that some unreachable uses might become reachable if
+// we spill in common dominator.
+std::pair<SetVectorType, SetVectorType>
+AMDGPUEarlyRegisterSpilling::collectReachableAndUnreachableUses(
```

Comment:
I think this is misuse of `reachable`. More accurate description would be `ReachableUsesExcludingDominated` or `NonDominatedReachableUses`. The dominated uses are reachable too.